### PR TITLE
Pako implementation to serialize Qualification document FE

### PR DESCRIPTION
**Pull request by ReinoRabbit**
- Implementing serialization and deserialization for employee qualifications
- Pako implementation for proof of qualification document

[Ticket on Azure board](https://dev.azure.com/retro-rabbit/HRIS/_sprints/taskboard/HRIS%20Team/HRIS/Sprint%2028?workitem=10442)
Video for reference
[doc serial.mp4](https://dev.azure.com/retro-rabbit/b91854dc-e8b7-46ec-b890-eab6439094a1/_apis/wit/attachments/574ca1e5-ebee-4237-bbbd-4725f5f61f09?fileName=doc%20serial.mp4&download=true)

This ticket will have FE & BE changes

### CODE REVIEW

File: src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts [hunk 1]
```diff
@@ -55,6 +55,10 @@ export class CareerSummaryQualificationsComponent {
     this.employeeQualificationsService.getEmployeeQualificationById(this.employeeProfile.employeeDetails.id as number).subscribe({
       next: async (data) => {
         this.sharedAccordionFunctionality.employeeQualification = data;
+        this.base64File = data.proofOfQualification;
+        console.log('Fetched qualifications:', data);
+        console.log('name of the fetched document', data.documentName);
+        console.log('This is the db proofOfQualification: ', this.base64File);
```
Comment (madwizzard) [inline src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts]:
```suggestion
```

File: src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts [hunk 2]
```diff
@@ -157,54 +160,25 @@ export class CareerSummaryQualificationsComponent {
...
-    for (let i = 0; i < byteString.length; i++) {
-      intArray[i] = byteString.charCodeAt(i);
-    }
-
-    const blob = new Blob([arrayBuffer], { type: 'application/pdf' });
-    const link = document.createElement('a');
-    link.href = window.URL.createObjectURL(blob);
-    link.download = this.fileName;
-    link.click();
+    console.log('downloadFile- Base64 file before decompression:', this.base64File);
```
Comment (madwizzard) [inline src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts]:
```suggestion
```

File: src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts [hunk 3]
```diff
@@ -157,54 +160,25 @@ export class CareerSummaryQualificationsComponent {
...
+    const decompressedFile = this.fileProcessingService.decompressFile(this.base64File);
+    console.log('downloadFile- Decompressed file:', decompressedFile);
```
Comment (madwizzard) [inline src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts]:
```suggestion
```

File: src/app/components/hris/employees/employee-profile/employee-profile.component.ts [hunk 4]
```diff
@@ -447,10 +451,11 @@ export class EmployeeProfileComponent implements OnChanges {
     const updatedEmployee = { ...this.employeeProfile };
     this.employeeProfileService.updateEmployeeProfile(updatedEmployee).subscribe({
       next: () => {
-        this.getEmployeeProfile();
-        this.snackBarService.showSnackbar("Updated", "snack-success");
+        console.log('Profile photo updated successfully.');
       },
-      error: (er) => this.snackBarService.showError(er),
+      error: (error) => {
+        console.error('Error updating profile photo:', error);
```
Comment (madwizzard) [inline src/app/components/hris/employees/employee-profile/employee-profile.component.ts]: Should this be a console.error ?

File: src/app/services/hris/file-processing.service.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FileProcessingService {
+
+  constructor() { }
+
+  validateFile(file: File): boolean {
+    console.log('Validating file:', file);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    return true;
+  }
+
+  convertFileToBase64(file: File): Promise<string> {
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onloadend = () => {
+        const base64File = reader.result as string;
+        if (base64File) {
+          console.log('convertFileToBase64- Converted file to base64:', base64File);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        console.error('downloadFile- Invalid base64 file data');
+        return;
+      }
+      const commaIndex = base64File.indexOf(',');
+      if (commaIndex === -1) {
+        console.error('downloadFile- Invalid base64 file data');
+        return;
+      }
+      const base64Data = base64File.slice(commaIndex + 1);
+      console.log('downloadFile- Base64 data before decoding:', base64Data);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+      for (let i = 0; i < byteString.length; i++) {
+        intArray[i] = byteString.charCodeAt(i);
+      }
+
+      const blob = new Blob([arrayBuffer], { type: 'application/octet-stream' });
+      const link = document.createElement('a');
+      link.href = window.URL.createObjectURL(blob);
+      link.download = fileName;
+      console.log('downloadFile- Triggering file download for:', fileName);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      link.click();
+    } catch (error) {
+      console.error('downloadFile- Error decoding base64 data or triggering download:', error);
+    }
+  }
+
+  compressFile(base64File: string): string {
+    try {
+      console.log('compressFile- Base64 data before compression:', base64File);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      const base64Data = base64File.split(',')[1];
+      const byteString = atob(base64Data);
+      const uint8Array = new Uint8Array(byteString.length);
+
+      for (let i = 0; i < byteString.length; i++) {
+        uint8Array[i] = byteString.charCodeAt(i);
+      }
+
+      const compressedData = pako.deflate(uint8Array);
+      console.log('compressFile- Compressed data:', compressedData);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+      const compressedBase64 = btoa(String.fromCharCode(...new Uint8Array(compressedData)));
+      console.log('compressFile- Compressed base64 data:', compressedBase64);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+      return compressedBase64;
+    } catch (error) {
+      console.error('compressFile- Error compressing file:', error);
+      throw error;
+    }
+  }
+
+  decompressFile(compressedBase64File: string): string {
+    try {
+      console.log('decompressFile- Compressed base64 data before decompression:', compressedBase64File);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+      const compressedBase64Data = compressedBase64File.split(',')[1] || compressedBase64File;
+      const compressedByteString = atob(compressedBase64Data);
+      const compressedUint8Array = new Uint8Array(compressedByteString.length);
+
+      for (let i = 0; i < compressedByteString.length; i++) {
+        compressedUint8Array[i] = compressedByteString.charCodeAt(i);
+      }
+
+      const decompressedUint8Array = pako.inflate(compressedUint8Array);
+      console.log('decompressFile- Decompressed data:', decompressedUint8Array);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/file-processing.service.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+      const decompressedBase64 = btoa(String.fromCharCode(...new Uint8Array(decompressedUint8Array)));
+      console.log('decompressFile- Decompressed base64 data:', decompressedBase64);
```
Comment (madwizzard) [inline src/app/services/hris/file-processing.service.ts]:
```suggestion
```

File: src/app/services/hris/image-processing.service.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  // Method to decompress and convert back to a base64 image
+  decompressImage(compressedData: Uint8Array): string {
+    const decompressed = pako.ungzip(compressedData);
+    const binaryString = String.fromCharCode.apply(null, Array.from(decompressed));
+    const base64Image = btoa(binaryString);
+
+    // Return the base64 image with the appropriate data URI prefix
+    return `data:image/png;base64,${base64Image}`;
+  }
+}
```
Comment (madwizzard) [inline src/app/services/hris/image-processing.service.ts]: Should we have comments in our code base ?

File: src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts [hunk 16]
```diff
@@ -99,44 +103,43 @@ export class CareerSummaryQualificationsComponent {
...
-      });
+        const saveQualification: EmployeeQualifications = {
+            id: this.sharedAccordionFunctionality.employeeQualification ? this.sharedAccordionFunctionality.employeeQualification.id : 0,
+            employeeId: this.employeeProfile.employeeDetails.id as number,
+            highestQualification: this.sharedAccordionFunctionality.employeeQualificationForm.get("highestQualification")?.value,
+            school: this.sharedAccordionFunctionality.employeeQualificationForm.get("school")?.value,
+            fieldOfStudy: this.sharedAccordionFunctionality.employeeQualificationForm.get("fieldOfStudy")?.value,
+            year: this.sharedAccordionFunctionality.employeeQualificationForm.get("year")?.value + "-01-01",
+            nqfLevel: this.sharedAccordionFunctionality.employeeQualificationForm.get("highestQualification")?.value,
+            proofOfQualification: this.fileProcessingService.compressFile(this.base64File),  // Ensure this is the latest base64 file
```
Comment (bseptember) [inline src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts]: Refactor: line 114 has comment that is not neccessary

File: src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts [hunk 17]
```diff
@@ -157,54 +160,25 @@ export class CareerSummaryQualificationsComponent {
...
+        this.fileUploaded = true;
+        const file = event.target.files[0];
+        this.fileName = file.name;
+        if (this.fileProcessingService.validateFile(file)) {
+            this.fileProcessingService.convertFileToBase64(file).then((base64File) => {
+                console.log('onFileChange- File selected, base64:', base64File);
+                this.base64File = base64File;
+                this.isDisabledDownload = false; 
+            }).catch((error) => {
+                console.error('Error converting file to base64:', error);
```
Comment (bseptember) [inline src/app/components/hris/employees/employee-profile/accordions/accordion-career-summary/accordion-career-summary-qualifications/accordion-career-summary-qualifications.component.ts]: Refactor: Perhaps throw a snackbar for the user

File: src/app/components/hris/employees/employee-profile/shared-accordion-functionality.ts [hunk 18]
```diff
@@ -82,7 +82,7 @@ export class SharedAccordionFunctionality {
     fieldOfStudy: "",
     year: "",
     nqfLevel: nqfLevels,
-    proofOfQualification: "",
+    proofOfQualification: '',
```
Comment (bseptember) [inline src/app/components/hris/employees/employee-profile/shared-accordion-functionality.ts]: Fix: shouldn't this be a [ ] ?

File: src/app/services/hris/image-processing.service.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
+import { Injectable } from '@angular/core';
```
Comment (bseptember) [inline src/app/services/hris/image-processing.service.ts]: Refactor: this file should not exist, there should be one file processing file

File: src/app/components/hris/employees/employee-profile/employee-profile.component.ts [hunk 20]
```diff
@@ -30,6 +30,8 @@ import { EmployeeTermination } from 'src/app/models/hris/employeeTermination.int
 import { Subscription } from 'rxjs';
 import { ClientService } from 'src/app/services/hris/client.service';
 import { SharedPropertyAccessService } from 'src/app/services/hris/shared-property-access.service';
+import * as pako from 'pako'; 
+import { ImageProcessingService } from 'src/app/services/hris/image-processing.service';
```
Comment (bseptember) [inline src/app/components/hris/employees/employee-profile/employee-profile.component.ts]: Refactor: pako import should not be used here, only file-processing service import should be used.

**Review by bseptember — CHANGES_REQUESTED** [review]

**Review by bseptember — APPROVED** [review]

**Review by madwizzard — APPROVED** [review]

**Review by MathildaShongwe15 — APPROVED** [review]

**Pull request merged**